---
title: "Class 16: Pathway Analysis from RNA-Seq Data"
author: "Larissa Rapadas"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## About our input data

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "***Differential analysis of gene regulation at transcript resolution with RNA-seq***". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

## Section 1. Differential Expression Analysis

First we have to load DESeq2 into our workspace
```{r}
library (DESeq2)
```

Next we load our data files in
```{r}
metaFile <- "data/GSE37704_metadata.csv"
countFile <- "data/GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)

# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

There are `r nrow(countData)` genes in this dataset and `r ncol(countData)` experiments (samples)...

Looks like there is a `length` column in out countData that we need to remove. 

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```
This looks better but there are lots of zero entries in there so let's get rid of them as we have no data for these.

```{r}
# Filter count data where you have 0 read count across all samples.
countData <- countData[rowSums(countData) != 0, ]
head(countData)
```
```{r}
nrow(countData)
```
## Running DESeq2

Now, lets setup the DESeqDataSet object required for the `DESeq()` function and then run the DESeq pipeline. This is again similar to our last days hands-on session.

```{r}
dds <- DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds <- DESeq(dds)
```

Let's see our results
```{r}
res <- results(dds, contrast=c("condition", "hoxa1_kd", "control_sirna"))
res
```
```{r}
summary(res)
```

## Let's plot the Volcano Plot

Summary plot of our results to date.

```{r}
plot( res$log2FoldChange, -log(res$padj) )
abline (h= -log(0.05), col = "red")
abline (v=c(-2,2), col="red")
```

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

## Adding gene annotation

Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.
```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```
```{r}
columns(org.Hs.eg.db)
```
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")
```
```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
```
```{r}
# res$name =   mapIds(org.Hs.eg.db,
#                     keys=row.names(res),
#                     keytype=ENSEMBL,
#                     column=___,
#                     multiVals="first")

```

## Pathway Analysis
```{r}
library(pathview)
```
```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```
Now, letâ€™s run the *gage* pathway analysis.

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
```

What is in our results?
```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

We use the KEGG Pathway ID from our top hit along with our `foldchange` vector to view our results as a more useful figure.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```
```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```
```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```
```{r}

```

